sensor: tdk 42688 driver

yperess opened the pull request:
Adds a driver for TDK InvenSense 42688 six axis IMU. Verified using the sensor shell sample app via:

sensor info
sensor get icm42688p@0
Collaborated on by Tom Burdick and Yuval Peress

Commit:
sensor: tdk 42688 driver

Adds a driver for TDK InvenSense 42688 six axis IMU. Verified using
the sensor shell sample app via:

- sensor info
- sensor get icm42688p@0

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>
Signed-off-by: Yuval Peress <peress@google.com>

teburd reviewed (dismissed):
I'm good with this, I have some patches I can PR once in.

MaureenHelm commented on dts/bindings/sensor/invensense,icm42688.yaml:
```diff
@@ -0,0 +1,98 @@
+include: [sensor-device.yaml, spi-device.yaml]
+
+properties:
+    friendly-name:
+      default: "ICM42688"
```
I don't think it makes sense to set a default friendly name in the binding; especially if that default is just the model name.

yperess replied:
Done

MaureenHelm commented on dts/bindings/sensor/invensense,icm42688.yaml:
```diff
@@ -0,0 +1,98 @@
+      description: |
+        Default frequency of accelerometer. (Unit - Hz)
+        Maps to ACCEL_ODR field in ACCEL_CONFIG0 setting
+        Power-on reset value is 1000.
```
Make the property optional and set the default to the power-on reset value

MaureenHelm commented on dts/bindings/sensor/invensense,icm42688.yaml:
```diff
@@ -0,0 +1,98 @@
+      description: |
+        Default frequency of gyroscope. (Unit - Hz)
+        Maps to GYRO_ODR field in GYRO_CONFIG0 setting
+        Power-on reset value is 1000.
```
Make the property optional and set the default to the power-on reset value

yperess replied:
Done

MaureenHelm commented on dts/bindings/sensor/invensense,icm42688.yaml:
```diff
@@ -0,0 +1,98 @@
+      description: |
+        Default full scale of accelerometer. (Unit - g)
+        Maps to ACCEL_FS_SEL field in ACCEL_CONFIG0 setting
+        Power-on reset value is 16
```
Make the property optional and set the default to the power-on reset value

yperess replied:
Done

MaureenHelm commented on dts/bindings/sensor/invensense,icm42688.yaml:
```diff
@@ -0,0 +1,98 @@
+      description: |
+        Default full scale of gyroscope. (Unit - DPS)
+        Maps to GYRO_FS_SEL field in GYRO_CONFIG0 setting
+        Power-on reset value is 2000
```
Make the property optional and set the default to the power-on reset value

yperess replied:
Done

MaureenHelm commented on samples/sensor/sensor_shell/prj.conf:
```diff
@@ -1,5 +1,7 @@
 CONFIG_SHELL=y
-CONFIG_BOOT_BANNER=n
+CONFIG_LOG=y
```
Agree they should be dropped

yperess replied:
Done

MaureenHelm reviewed (changes requested):
Please add the sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI

keith-zephyr commented on drivers/sensor/icm42688/icm42688.c:
```diff
@@ -0,0 +1,213 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ICM42688_LL, CONFIG_SENSOR_LOG_LEVEL);
+
+void icm42688_lock(const struct device *dev)
```
Why do you have an empty lock function?

yperess replied:
It was used in Tom's additional changes. I'll remove for now until it's needed.

keith-zephyr commented on drivers/sensor/icm42688/icm42688_core.c:
```diff
@@ -0,0 +1,201 @@
+	return -ENOTSUP;
+}
+
+static const struct sensor_driver_api icm42688_driver_api = {
+	.sample_fetch = icm42688_sample_fetch,
+	.channel_get = icm42688_channel_get,
+	.attr_set = icm42688_attr_set,
+	.attr_get = icm42688_attr_get,
+};
+
+int icm42688_init(const struct device *dev)
```
Nit - I'd prefer that the driver implementation is provided by the module matching the sensor name.  So this file would be `icm42688.c` and the file with the helper routines `icm42688_reset()` and `icm42688_configure()` would be called `icm42688_core.c` or maybe `icm42688_support.c`.

@MaureenHelm what your thoughts?

yperess replied:
I ended with `icm42688_common.c` and `icm42688.c`, does that work?

yperess commented on boards/arm/tdk_robokit1/tdk_robokit1_defconfig:
```diff
@@ -12,6 +12,8 @@ CONFIG_SERIAL=y
 CONFIG_BOARD_TDK_ROBOKIT1=y
 CONFIG_WDT_DISABLE_AT_BOOT=y
 CONFIG_BUILD_OUTPUT_HEX=y
+CONFIG_SPI=y
+CONFIG_SPI_SAM=y
```
Done

yperess commented on drivers/sensor/icm42688/Kconfig:
```diff
@@ -0,0 +1,14 @@
+
+config ICM42688
+	bool "ICM42688 Six-Axis Motion Tracking Device"
+	default $(dt_compat_enabled,$(DT_COMPAT_ICM42688))
+	depends on SPI
```
Thanks, that's cool, didn't know about the `DT_HAS_..._ENABLED` macro

yperess commented on drivers/sensor/icm42688/icm42688.h:
```diff
@@ -0,0 +1,448 @@
+		    sensitivity;
+}
+
+/**
+ * @brief The value of gravitational constant in micro m/s^2.
+ * TODO this is redefined because sensor v2 redefines SENSOR_G
+ */
+#define ICM42688_G 9806650LL
+
+/**
+ * @brief The value of constant PI in micros.
+ * TODO this is redefined because sensor v2 redefines SENSOR_PI
+ */
+#define ICM42688_PI 3141592LL
```
Done

yperess commented:
> Please add the sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI

Done

Commit:
tests: clean up build_all/sensor

The test at tests/drivers/build_all/sensor wasn't passing since a lot
of nodes in both the i2c and spi had conflicting addresses. Fix that by
making the addresses sequential.

Signed-off-by: Yuval Peress <peress@google.com>

keith-zephyr commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -41,90 +41,90 @@ test_i2c_iaqcore: iaqcore@5 {
 };
 
-test_i2c_bme280: bme280@7 {
+test_i2c_bme280: bme280@6 {
```
Nit - add a comment to the file header indicating that the I2C peripheral addresses should be sequential.

yperess replied:
Done

keith-zephyr commented on tests/drivers/build_all/sensor/spi.dtsi:
```diff
@@ -26,9 +26,9 @@ test_spi_apa102: apa102@2 {
 };
 
-test_spi_rf2xx: rf2xx@4 {
+test_spi_rf2xx: rf2xx@3 {
```
Same here - document that the SPI addresses should be sequential.

yperess replied:
Done

keith-zephyr commented on tests/drivers/build_all/sensor/spi.dtsi:
```diff
@@ -318,8 +318,20 @@ test_spi_icm42670: icm42670@3c {
 	spi-max-frequency = <0>;
 };
+
+/* Requires the SPI to be atml,sam-spi
```
I assume you meant to submit this as a separate commit?

yperess replied:
I left it in there so Tom & Maureen could see what I tried. It's been split up into a separate commit now that it works.

Commit:
test: add icm42688 to build_all/sensor test

Add a node for the icm42688 so that it builds as a part of the CI.

Signed-off-by: Yuval Peress <peress@google.com>

yperess commented:
> > Please add the sensor to `tests/drivers/build_all/sensor` to ensure it gets built in CI
> 
> @MaureenHelm I tried this but ran into a few issues (see the second commit in this PR)
> 
> 1. The test wasn't passing due to conflicting addresses (I fixed that)
> 2. The driver requires the SPI_SAM (@teburd can you confirm?) which can't be built for native_posix, so I can't seem to add it to this test without major refactoring.

Discussed with @teburd offline and there's no hard requirement for the SPI_SAM, CI should pass now

MaureenHelm commented on drivers/sensor/icm42688/icm42688.h:
```diff
@@ -0,0 +1,422 @@
+		raw[i] = sys_le16_to_cpu((data[i * 2] << 8) | data[i * 2 + 1]);
+	}
+
+	printk("temp raw %d, accel x %d, y %d, z %d, gyro x %d, y %d, z %d\n", raw[0], raw[1],
+	       raw[2], raw[3], raw[4], raw[5], raw[6]);
```
`LOG_DBG`

yperess replied:
Done

MaureenHelm reviewed (changes requested):
Minor nit, but otherwise looks good. Also needs a rebase
